### ISSUE TITLE
Global/System Hook System

### ISSUE THREAD

**Issue by rbroemeling** [Shopify/shipit-engine#841]
Hi Shopify,

We want to modify our local ShipIt install to track deployment metrics (how long are deploys taking? how many deploys are happening? Etc.) We track all metrics via DataDog, so what we want to do is report this information directly to DataDog somehow. We have a local dogstatsd instance that is available, so it would be straightforward to do this if we could figure out how to execute some code at specific points in the ShipIt process ("deploy begun", and "deploy complete", for instance).

I realize that we can do this with `post`/`pre` overrides in the ShipIt configuration of each repository, but that's pretty far from ideal because it means that we need to set that up as a part of every single repository, and that someone has to notice if it is missed, etc.

We'd much rather handle this at the system/ShipIt level, where every stack gets it by default, without any developer pain or requirement to ask for it or implement it.

I am further aware that ShipIt has web hook support (which is waiting for a UI from #364), and I briefly looked to see if it was flexible enough to actually post metrics to DataDog's HTTP metric submission API directly. It doesn't look like there is any ability to template or control the body of the sent data, though, so that doesn't appear viable.

So we're left in a situation where we need to implement an interstitial web daemon that just waits for hooks from ShipIt and then "translates" them into DataDog stats (probably submitting them to the local dogstatsd instance). That isn't really a desirable infrastructure from our point-of-view, however.

ShipIt doesn't currently seem to have a global hook script system (i.e., for each event run all of the executable scripts in a given directory).

This seems like it would be a significant win in terms of flexibility (it'd give a huge amount of flexibility to ShipIt, as people would be able to customize behaviours for the entire ShipIt system by just dropping scripts into this set of directories) -- is this a feature request that the ShipIt team at Shopify would be interested in, or is it undesirable for some reason?

Example: what I'm looking for is something similar to the Buildkite Agent hooks system (https://buildkite.com/docs/agent/v3/hooks).

**Comment by rbroemeling** [comment]
I've been looking at a possible implementation of this.

One possible way would be to add a `LocalHookRunner` background job that takes an event, maps the event to a code directory, and then executes all of the scripts inside that directory in a deterministic order. Then, in `Shipit::Task` around line 285, add emission of a `LocalHookRunner` immediately after `Hook` is emitted, possibly with a guard to only emit the job if-and-only-if the directory exists (to prevent emitting a bunch of "no-op" background jobs).

Does that mechanism make sense? I'm willing to take on some of this work and to prepare a PR, if it would be welcomed.

**Comment by casperisfine** [comment]
Hi. What we're currently doing for this is that we monkey patch some classes from our internal installation. It's clearly not very clean so I definitely wouldn't mind some proper API for that.

> One possible way would be to add a LocalHookRunner background job that takes an event

I don't think a background job is needed. Since Shipit is an engine, users can write initializers etc in the host app.

IMO a simple event listener API would be more than enough, either using [ActiveSupport::Notifications](https://api.rubyonrails.org/classes/ActiveSupport/Notifications.html):

```ruby
ActiveSupport::Notifications.subscribe('shipit.task.perform') do |name, start, finish, id, payload|
  task = payload[:task]
  tags = %W(
    type:#{task.class.name.demodulize.underscore}
    by:#{task.author.login}
    stack:#{task.stack.to_param}
    status:#{task.status}
  )
  StatsD.increment('task.perform', tags: tags)
  StatsD.measure('task.duration', duration, tags: tags)
end
```

or something even simpler.

**Comment by doodzik** [comment]
closed by #865
